File: apps/backend-hash/products/src/support/Configuration.js
const nconf = require('nconf');

class Configuration {
  constructor() {
    nconf.env();
  }

  app() {
    return {
      name: () => nconf.get('APP_NAME'),
      env: () => nconf.get('APP_ENV'),
    };
  }

  log() {
    return {
      level: () => nconf.get('LOG_LEVEL') || 'ERROR',
    };
  }

  protos() {
    return {
      v1: () => 'api.v1.proto',
    };
  }

  http() {
    return {
      port: () => nconf.get('HTTP_PORT'),
    };
  }

  grpc() {
    return {
      root: () => './proto',
      address: () => `0.0.0.0:${nconf.get('GRPC_PORT')}`,
      proto: name => `${this.grpc().root()}/${name}`,
    };
  }

  discounts() {
    return {
      root: () => '../discounts/src/main/protobuf',
      address: () => `${nconf.get('DISCOUNTS_ADDRESS')}:${nconf.get('DISCOUNTS_PORT')}`,
      proto: () => `${this.discounts().root()}/api.v1.proto`,
    };
  }

  database() {
    return {
      host: () => nconf.get('DB_HOST'),
      user: () => nconf.get('DB_USERNAME'),
      pass: () => nconf.get('DB_PASSWORD'),
      name: () => nconf.get('DB_DATABASE'),
    };
  }
}

module.exports.config = new Configuration();

File: apps/backend-hash/discounts/src/main/scala/com/pasviegas/discounts/support/Configuration.scala
package com.pasviegas.discounts.support

import com.typesafe.config.{Config, ConfigFactory}

class Configuration {
  import net.ceedubs.ficus.Ficus._

  private val config: Config = ConfigFactory.load()

  def app: AppConfig.type = AppConfig

  def log: LogConfig.type = LogConfig

  def grpc: GrpcConfig.type = GrpcConfig

  def database: DatabaseConfig.type = DatabaseConfig

  object AppConfig {
    def name: String = config.as[String]("app.name")
    def env: String  = config.as[String]("app.env")
  }

  object LogConfig {
    def level: String = config.as[String]("log.level")
  }

  object GrpcConfig {
    def port: Int = config.as[Int]("grpc.port")
  }

  object DatabaseConfig {
    def host: String = config.as[String]("database.host")
    def name: String = config.as[String]("database.name")
    def user: String = config.as[String]("database.user")
    def pass: String = config.as[String]("database.pass")
  }
}
